Source/WebCore/rendering/BackgroundPainter.cpp

```diff
@@ -327,18 +331,32 @@ void BackgroundPainter::paintFillLayer(const Color& color, const FillLayer& bgLa
     auto backgroundClipOuterLayerScope = TransparencyLayerScope(context, 1, false);
     auto backgroundClipInnerLayerScope = TransparencyLayerScope(context, 1, false);
 
-    auto setupMaskingBackgroundClip = [&](const LayoutRect& borderRect, const std::function<void(const LayoutRect&, const FloatRect&)>& paintFunction) {
+    auto setupMaskingBackgroundClip = [&](const LayoutRect& borderRect, const std::function<void(GraphicsContext& context, const LayoutRect&, const FloatRect&)>& paintFunction) {
         auto transparencyLayerBounds = snapRectToDevicePixels(rect, deviceScaleFactor);
         transparencyLayerBounds.intersect(snapRectToDevicePixels(m_paintInfo.rect, deviceScaleFactor));
         transparencyLayerBounds.inflate(1);
 
         backgroundClipStateSaver.save();
         context.clip(transparencyLayerBounds);
-
         backgroundClipOuterLayerScope.beginLayer(1);
-        paintFunction(borderRect, transparencyLayerBounds);
 
-        context.setCompositeOperation(CompositeOperator::SourceIn);
+        if (context.renderingMode() == RenderingMode::PDFDocument) {
+            // CG doesn't support some compositing operations when printing, so clip using an imageBuffer instead.
```

Comment:
Sad to see mention of CG here, and also workarounds for CG here.

Reply:
Can it at least be a "GraphicsContext::needs/supportsXYZ" and move the mention of CG in there?